doc: Design for certificate pruning

rcritten opened the pull request:
This describes how the certificate pruning capability of PKI introduced in v11.3.0 will be integrated into IPA, primarily for ACME.

Related: https://pagure.io/freeipa/issue/9294

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+* day of week is within 0-6
+
+No validation of setting February 31st will be done. That will be left to PKI. Buyer beware.
+
```
What happens if the admin calls `ipa-acme-manage pruning --cron=...` but the job is disabled?

rcritten replied:
The configuration is stored but the job won't be run until it is enabled.

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.requestSearchTimeLimit 0`
+
+A value of 0 for searchtimelimit is unlimited.
```
I assume that the searchtimelimit and searchsizelimit are client-side values but the results can also be limited by server-side limits. Which user is ultimately performing the search on the LDAP server? Does pki use uid=pkidbuser? If that's the case it would be worth mentioning that the limits provided through `ipa-acme-manage pruning --requestsearchsizelimit=VALUE --requestsearchtimelimit=VALUE` cannot override the limits for this user.

rcritten replied:
pki binds with internaldb.ldapauth.bindDN in CS.cfg which is Directory Manager, so no limits.

flo-renaud replied:
On my instance, I have by default:
```
internaldb.ldapauth.authtype=SslClientAuth
internaldb.ldapauth.bindDN=cn=Directory Manager
internaldb.ldapauth.bindPWPrompt=internaldb
internaldb.ldapauth.clientCertNickname=subsystemCert cert-pki-ca
internaldb.ldapconn.host=server.ipa.test
internaldb.ldapconn.port=636
internaldb.ldapconn.secureConn=true
```
and all the connections between pki and the LDAP server are authenticated with `subsystemCert cert-pki-ca`, linked to uid=pkidbuser,ou=people,o=ipaca.

rcritten replied:
internaldb.ldapauth.bindDN is what @edewata told me. It may be because its an internal task.

edewata replied:
I think it depends on the `internaldb.ldapauth.authtype`. If it's set to `BasicAuth` (which is the default) it will use `internaldb.ldapauth.bindDN`. If it's set to `SslClientAuth` it will use the user mapped to the cert specified in `internaldb.ldapauth.clientCertNickname`.

edewata replied:
So I don't think the `internaldb.ldapauth.bindDN` and `internaldb.ldapauth.bindPWPrompt` are needed in IPA case.

rcritten replied:
Thanks. I added a mention of the limits and a pointer to the 389-ds tuning documentation.

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
````diff
@@ -0,0 +1,282 @@
+Output may look something like:
+
+```
+# ipa-cacert-manage --config-show
````
ipa-cacert-manage or ipa-acme-manage?

rcritten replied:
Nice catch, thanks, I'll fix it up.

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+| ipa-acme-manage | --certsearchsizelimit=1000 --certsearchtimelimit=0 |
+| ipa-acme-manage | --requestretention=30 --requestretentionunit=day |
+| ipa-acme-manage | --requestsearchsizelimit=1000 --requestsearchtimelimit=0 |
+| ipa-acme-manage | --config-show |
```
replace `ipa-acme-manage` with `ipa-acme-manage pruning` in the table

rcritten replied:
Right

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+| ipa-acme-manage | --requestretention=30 --requestretentionunit=day |
+| ipa-acme-manage | --requestsearchsizelimit=1000 --requestsearchtimelimit=0 |
+| ipa-acme-manage | --config-show |
+
```
It's worth mentioning that ipa-acme-manage can be run by root only.

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+* moving time forward 31 days
+* manually running the job
+* validating that the certificates are removed
+
```
In order to test the configuration, create a large number of certificates/requests and set the search limit to a low value, then ensure that the number of deleted certs is equal to the search limit.

rcritten replied:
Added.

flo-renaud reviewed (commented):
Hi @rcritten 
please find inline comments/questions.

rcritten commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,282 @@
+
+In addition to the automated schedule it is possible to manually run the pruning job.
+
+The configuration is not replicated and needs to be configured on each IPA server by the administrator.
```
The assumption in my head, which I didn't write down, was one or two servers to be configured with pruning depending on size of deployment.

This is a similar problem to CRL generation except the only real consequence, I think, is extra replication. These are delete operations so I don't think there will be any issues other than potentially unnecessary replication overhead if configured on multiple servers. I suppose we could lump this in with CRL generation and specify one server to run it on. The single-point-of-failure always worries me.

Agreed that different configuration could be odd, and perhaps difficult to troubleshoot, but we couldn't prevent a user from doing this. If we enforce that the config is only set on the "pruning server" then if that server dies, how do they move it? I'm fine relying on the responsibility of the users.

flo-renaud replied:
I'm fine with the recommendation (one or two servers configured for pruning) and I think it should be explained in the design, with the rationale.

rcritten replied:
I added a section on this.

rcritten commented:
I'm inclined to remove the requirement that ACME be enabled because while this is primarily for ACME, it isn't necessarily a prerequisite.

Other than that, are there any additional comments?

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,297 @@
+* Additional stress on the LDAP server searching for expirations
+* Unnecessary replication load deleting the same entries on multiple servers
+
+While enabling this on a signle server represents a single-point-of-failure there should be no catastrohpic consequences other than expired certificates and requests potentially building up. This can be cleared by enabling pruning on a different server. Depending on the size of the backlog this could take a couple of executions to catch up.
```
Typo catastro**hp**ic

rcritten replied:
fixed.

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,297 @@
+
+### Enabling pruning
+
+`$ ipa-acme-manage pruning --enable=TRUE`
```
Nitpick but ipa-acme-manage requires to be executed as root, I would replace the `$` prompt with a `#` prompt, or with `$ sudo ipa-acme-manage`. This comment applies to all the examples with $ ipa-acme-manage

rcritten replied:
Fixed

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
```diff
@@ -0,0 +1,297 @@
+
+### Configuring job execution time
+
+The CA provides a cron-like interface for scheduling jobs. To configure the job to run at midnight on the first of every month the PKI equivlant command-line is:
```
Typo equi**vl**ant

rcritten replied:
Fixed

flo-renaud commented on doc/designs/expired_certificate_pruning.md:
````diff
@@ -0,0 +1,297 @@
+
+Output may look something like:
+
+```
````
You can use ```console to display the output with a fancier look (otherwise the text is treated as code and # as a comment)

rcritten replied:
Added. Thanks, that does look a lot better.

flo-renaud reviewed (commented):
Hi @rcritten 
I only spotted a few typos or style nitpicks.

rcritten commented:
I used ghostwriter which has a spell checker and fixed a few more typos here and there.

I updated the Testing section to drop creating a new profile. If we're moving time anyway then 90 vs 30 days is not a big deal.

I included that the pki-server configuration options are case-sensitive and not applied until the CA is restarted.

flo-renaud commented:
@rcritten thanks for the update, ack for me when ci completes.

rcritten commented:
master:

* 5d9f59038db8c8b133f3d75ce7d98324daf5b403 doc: Design for certificate pruning
